mempool/cat/reactor.go

```diff
@@ -0,0 +1,448 @@
+package cat
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+
+	cfg "github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/crypto/tmhash"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/mempool"
+	"github.com/tendermint/tendermint/p2p"
+	protomem "github.com/tendermint/tendermint/proto/tendermint/mempool"
+	"github.com/tendermint/tendermint/types"
+)
+
+const (
+	// default duration to wait before considering a peer non-responsive
+	// and searching for the tx from a new peer
+	defaultGossipDelay = 200 * time.Millisecond
+
+	// Content Addressable Tx Pool gossips state based messages (SeenTx and WantTx) on a separate channel
+	// for cross compatibility
+	MempoolStateChannel = byte(0x31)
+)
+
+// Reactor handles mempool tx broadcasting logic amongst peers. For the main
+// logic behind the protocol, refer to `ReceiveEnvelope` or to the english
+// spec under /.spec.md
+type Reactor struct {
+	p2p.BaseReactor
+	opts     *ReactorOptions
+	mempool  *TxPool
+	ids      *mempoolIDs
+	requests *requestScheduler
+}
+
+type ReactorOptions struct {
+	// ListenOnly means that the node will never broadcast any of the transactions that
+	// it receives. This is useful for keeping transactions private
+	ListenOnly bool
```

Comment:
What would be considered as a private transaction? can you please give an example?

Reply:
A private transaction is a transaction that no other node is aware of until it is first proposed in a block. People may want private transaction to avoid something like front running